[WFLY-14074] Normalize principial propagation/injection across elytron…

skyllarr opened the pull request:
… and legacy

https://issues.redhat.com/browse/WFLY-14074

OndrejKotek commented on ejb3/src/main/resources/schema/wildfly-ejb3_8_0.xsd:
```diff
@@ -478,6 +478,13 @@
+        <xs:attribute name="legacy-compliant-principal-propagation" type="xs:boolean" use="optional" default="true">
+            <xs:annotation>
+                <xs:documentation>
+                    Option to switch between legacy compliant principal propagation or Elytron principal propagation.
```
I suppose a simple explanation of the diference would be helpful.

Skyllarr replied:
@OndrejKotek  fixed.

fjuma commented on ejb3/src/main/java/org/jboss/as/ejb3/subsystem/ApplicationSecurityDomainDefinition.java:
```diff
@@ -96,7 +96,13 @@
-    private static final AttributeDefinition[] ATTRIBUTES = new AttributeDefinition[] { SECURITY_DOMAIN, ENABLE_JACC };
+    static final SimpleAttributeDefinition LEGACY_COMPLIANT_PRINCIPAL_PROPAGATION = new SimpleAttributeDefinitionBuilder(EJB3SubsystemModel.LEGACY_COMPLIANT_PRINCIPAL_PROPAGATION, ModelType.BOOLEAN, true)
+            .setDefaultValue(ModelNode.TRUE)
+            .setMinSize(1)
```
Looks like `setMinSize` could be removed since this attribute is a boolean value.

Skyllarr replied:
@fjuma fixed

chengfang commented:
wildfly-ejb3 schema was already at version 8. So you will need to create a new schema file, like wildfly-ejb3_9_0.xsd, to contain your additions.  If there is other features later within the current release that will add to the schema, they can modify on top of your version of ejb3 schema.

skyllarr committed:
[WFLY-14074] Update EJB3 schema from 8.0 to 9.0

skyllarr committed:
[WFLY-14074] Normalize principial propagation/injection across elytron and legacy

Skyllarr commented:
@fjuma @chengfang I added another commit with update of EJB3 schema from 8.0 to 9.0. Thanks!

Skyllarr commented:
I just added community documentation to this PR.

fjuma commented on docs/src/main/asciidoc/_elytron/Elytron_and_Java_Authorization_Contract_for_Containers-JACC.adoc:
```diff
@@ -87,3 +87,5 @@ to enable JACC you can execute a command as follows:
 [standalone@localhost:9990 /] /subsystem=ejb3/application-security-domain=my-security-domain:write-attribute(name=enable-jacc,value=true)
 ----
+
+The `application-security-domain` resource of the EJB3 subsystem also has an option `legacy-compliant-principal-propagation`. If this is set to true and there is no incoming run-as identity, then the principal of the local unsecured bean is the current authenticated identity. This was the case for legacy PicketBox security. If this attribute is set to false, the behaviour will comply with the Ely ...
```
The text here looks good. Maybe instead of adding it in this document, we could add it to the EJB3 section here:

https://github.com/wildfly/wildfly/blob/master/docs/src/main/asciidoc/_elytron/Using_WildFly_Elytron_with_WildFly.adoc

Skyllarr replied:
@fjuma Fixed. Thanks

fjuma commented:
@chengfang Please review.

Skyllarr commented:
> wildfly-ejb3 schema was already at version 8. So you will need to create a new schema file, like wildfly-ejb3_9_0.xsd, to contain your additions. If there is other features later within the current release that will add to the schema, they can modify on top of your version of ejb3 schema.

@chengfang After I updated ejb3 schema to 9.0 I started to get mixed domain failures and I am not yet able to figure out what is wrong. Do you maybe have any idea what could be wrong with my schema update commit? Thank you

fjuma commented on ejb3/src/main/java/org/jboss/as/ejb3/subsystem/EJB3Extension.java:
```diff
@@ -119,5 +120,6 @@ public void initializeParsers(ExtensionParsingContext context) {
         context.setSubsystemXmlMapping(SUBSYSTEM_NAME, NAMESPACE_6_0, EJB3Subsystem60Parser::new);
         context.setSubsystemXmlMapping(SUBSYSTEM_NAME, NAMESPACE_7_0, EJB3Subsystem70Parser::new);
         context.setSubsystemXmlMapping(SUBSYSTEM_NAME, NAMESPACE_8_0, EJB3Subsystem80Parser::new);
+        context.setSubsystemXmlMapping(SUBSYSTEM_NAME, NAMESPACE_9_0, EJB3Subsystem90Parser::new);
```
@Skyllarr Just noticed that it looks like there is also a `CURRENT_MODEL_VERSION` in this file that needs to be updated.

Skyllarr replied:
@fjuma Fixed

chengfang commented on ejb3/src/main/java/org/jboss/as/ejb3/subsystem/ApplicationSecurityDomainService.java:
```diff
@@ -134,6 +138,10 @@ public boolean isEnableJacc() {
+        public boolean isLegacyCompliantPrincipalPropagation() {
+            return legacyCompliantPrincipalPropagation;
+        }
+
```
looks like this method is not used.

Skyllarr replied:
Fixed

Skyllarr replied:
Fixed

chengfang commented on ejb3/src/main/java/org/jboss/as/ejb3/subsystem/ApplicationSecurityDomainService.java:
```diff
@@ -120,10 +122,12 @@ public void cancel() {
+        private final boolean legacyCompliantPrincipalPropagation;
 
-        public ApplicationSecurityDomain(final SecurityDomain securityDomain, boolean enableJacc) {
+        public ApplicationSecurityDomain(final SecurityDomain securityDomain, boolean enableJacc, boolean legacyCompliantPrincipalPropagation) {
```
this flag `legacyCompliantPrincipalPropagation` is passed from the enclosing class. Maybe we can just omit it and access the flag in the outer class directly.

Skyllarr replied:
Fixed

chengfang commented on ejb3/src/main/java/org/jboss/as/ejb3/subsystem/EJB3Subsystem50Parser.java:
```diff
@@ -110,7 +110,7 @@ private void parseApplicationSecurityDomains(final XMLExtendedStreamReader reade
     }
 
-    private void parseApplicationSecurityDomain(final XMLExtendedStreamReader reader, final List<ModelNode> operations) throws XMLStreamException {
+    protected void parseApplicationSecurityDomain(final XMLExtendedStreamReader reader, final List<ModelNode> operations) throws XMLStreamException {
```
looks like only this method is overridden in subclass, so we probably don't need `protected` access modifier for other methods.

Skyllarr replied:
Fixed

chengfang commented on ejb3/src/main/resources/org/jboss/as/ejb3/subsystem/LocalDescriptions.properties:
```diff
@@ -405,6 +405,7 @@ application-security-domain.security-domain=The Elytron security domain to be us
 application-security-domain.referencing-deployments=The deployments currently referencing this mapping
 application-security-domain.enable-jacc=Enable authorization using JACC
 application-security-domain.http-authentication-factory=The HTTP authentication factory to use for HTTP based invocations
+application-security-domain.legacy-compliant-principal-propagation= Option to switch between legacy compliant principal propagation or Elytron principal propagation
```
there is a leading whitespace, which most likely won't matter, but just to follow the existing formatting

Skyllarr replied:
Fixed

chengfang commented:
> > wildfly-ejb3 schema was already at version 8. So you will need to create a new schema file, like wildfly-ejb3_9_0.xsd, to contain your additions. If there is other features later within the current release that will add to the schema, they can modify on top of your version of ejb3 schema.
> 
> @chengfang After I updated ejb3 schema to 9.0 I started to get mixed domain failures and I am not yet able to figure out what is wrong. Do you maybe have any idea what could be wrong with my schema update commit? Thank you

@Skyllarr are you still seeing mixed domain failures? It could be the new attribute is not properly handled by `EJBTransformers`. Which tests failed?

Skyllarr commented:
> > > wildfly-ejb3 schema was already at version 8. So you will need to create a new schema file, like wildfly-ejb3_9_0.xsd, to contain your additions. If there is other features later within the current release that will add to the schema, they can modify on top of your version of ejb3 schema.
> > 
> > 
> > @chengfang After I updated ejb3 schema to 9.0 I started to get mixed domain failures and I am not yet able to figure out what is wrong. Do you maybe have any idea what could be wrong with my schema update commit? Thank you
> 
> @Skyllarr are you still seeing mixed domain failures? It could be the new attribute is not properly handled by `EJBTransformers`. Which tests failed?

The following classes have failing tests:

```
org.jboss.as.test.integration.domain.mixed.eap720.MixedDomain720TestSuite
org.jboss.as.test.integration.domain.mixed.eap720.MixedDomainOverlay720TestSuite
org.jboss.as.test.integration.domain.mixed.eap730.MixedDomain730TestSuite
org.jboss.as.test.integration.domain.mixed.eap730.MixedDomainOverlay730TestSuite
```

error is mostly "WFLYDC0022: Operation failed or was rolled back on all servers." It might be because of EJBTransformers, however the test `Ejb3TransformersTestCase` is passing and I can't figure out what is wrong with the existing transformers.

chengfang commented:
I'm also seeing the mixed-domain failure in my local testing. I'll take a closer look. In the meantimem, you may want to ask in Zulip chat about "WFLYDC0022: Operation failed or was rolled back on all servers."

Skyllarr commented:
> I'm also seeing the mixed-domain failure in my local testing. I'll take a closer look. In the meantimem, you may want to ask in Zulip chat about "WFLYDC0022: Operation failed or was rolled back on all servers."

@chengfang  Thank you a lot, I will investigate further as well. And just FYI, only the single commit titled `Update EJB3 schema from 8.0 to 9.0` from this PR is sufficient to make these tests fail.

Skyllarr commented:
@chengfang The logs of the slave `server-one` have the following error:

```
ERROR [org.jboss.as.server] (ServerService Thread Pool -- 76) WFLYSRV0021: Deploy of deployment "test.war" was rolled back with the following failure message: undefined
```

and

```
WFLYCTL0184:    New missing/unsatisfied dependencies:
      service jboss.remoting.remotingConnectorInfoService.undefined (missing) dependents: [service org.wildfly.clustering.cache.registry-entry.ejb.client-mappings] 
WFLYCTL0448: 4 additional services are down due to their dependencies being missing or failed
```

Unfortunately I still can't find anything in my schema update that is making this error happen.

pferraro commented on ejb3/src/main/java/org/jboss/as/ejb3/subsystem/EJBTransformers.java:
```diff
@@ -303,7 +305,17 @@ private static void registerTransformers_7_0_0(ResourceTransformationDescription
+        subsystemBuilder.addChildResource(PathElement.pathElement(EJB3SubsystemModel.APPLICATION_SECURITY_DOMAIN))
+                .getAttributeBuilder()
+                .setDiscard(DiscardAttributeChecker.UNDEFINED, EJB3SubsystemModel.LEGACY_COMPLIANT_PRINCIPAL_PROPAGATION)
+                .addRejectCheck(RejectAttributeChecker.DEFINED, EJB3SubsystemModel.LEGACY_COMPLIANT_PRINCIPAL_PROPAGATION)
```
Correct me if I'm wrong, but it looks like you can also also discard this attribute if set to the default value (true), as this is consistent with the legacy implied behavior.

Skyllarr replied:
@pferraro yes true. I updated the PR

chengfang replied:
`connectors` and `remote-http-connection` should already be handled by `registerTransformers_7_0_0`, with the chained transformers. Why do we need to duplicate the same in `registerTransformers_8_0_0`?

Skyllarr replied:
@chengfang That was my initial thought as well, however after I changed it the MixedDomainDeployment730TestCase started to pass for me locally.

Skyllarr replied:
@chengfang Although looking at the CI now it seems I have messed up my local configuration and tests are still failing. So something else is wrong.

Skyllarr replied:
@chengfang @pferraro Actually I was looking at the wrong CI result before. Copying contents of registerTransformers_7_0_0 to registerTransformers_8_0_0 did help and all tests passed now.

Skyllarr replied:
@chengfang please let me know if you agree with this change.

chengfang replied:
let me check with @rachmatowicz , who did the `connectors` refactoring in ejb3 8.0.0 version. Richard, can you check why we need to duplicate the `connectors` part in `registerTransformers_8_0_0()`?

chengfang replied:
We can look into this EJB transformer issue as a separate one without blocking this feature, since this is not core feature of this PR.

Skyllarr replied:
@chengfang that might be good. But this feature require merging it

pferraro replied:
@chengfang There may have been an issue with the connectors transformers - as it look like there were no EAP 7.3 transformation tests until now.  I agree that this does not need to block this PR, but we should fix this before the WF 23 final release.

pferraro reviewed (changes requested):
Ejb3TransformersTestCase is missing transformer tests for EAP 7.3.

Skyllarr commented:
> Ejb3TransformersTestCase is missing transformer tests for EAP 7.3.

Thank you for review! I added this test case, hopefully the versions there are correct.

pferraro commented:
> @pferraro Thank you for review! I added this test case, hopefully the versions there are correct.

Skyllarr commented:
> > @pferraro Thank you for review! I added this test case, hopefully the versions there are correct.
> 
> We also need an EAP 7.3 version of the rejection style transformer test, i.e. `testRejectionsEAP730()`

@pferraro I added this test. Thank you.

Skyllarr commented:
@OndrejKotek let me know whether pre-check can be granted to this PR. Thanks!

OndrejKotek commented:
> @OndrejKotek let me know whether pre-check can be granted to this PR. Thanks!

I have executed WildFly and EJB test suites on our CI, no regressions found, pre-checked.

Skyllarr commented:
@OndrejKotek  @chengfang Thank you both.

bstansberry commented:
@fjuma @pferraro Please complete your review so we can retest this one and get it in. Thanks!

bstansberry commented:
/retest
